python/hail/dataset.py

```diff
@@ -4980,38 +4987,98 @@ def tdt(self, pedigree, root='va.tdt'):
+         - **pval** (*Float64*) -- p-value.
 
         :param pedigree: Sample pedigree.
         :type pedigree: :class:`~hail.representation.Pedigree`
 
-        :param root: Variant annotation root to store TDT result.
-
-        :return: Variant dataset with TDT association results added to variant annotations.
-        :rtype: :py:class:`.VariantDataset`
+        :return: Key table with TDT association results.
+        :rtype: :py:class:`.KeyTable`
         """
 
-        jvds = self._jvdf.tdt(pedigree._jrep, root)
-        return VariantDataset(self.hc, jvds)
+        trio_matrix = self.trio_matrix(pedigree, complete_trios=True)
+        trio_matrix = trio_matrix.filter_samples_expr('isDefined(sa.isFemale)')
+
+        hom_ref = 0
+        het = 1
+        hom_var = 2
+
+        auto = 2
+        hemiX = 1
+
+        t = TDict(TStruct(['kid', 'dad', 'mom', 'ploidy'], [TInt32(), TInt32(), TInt32(), TInt32()]), TArray(TInt32()))
+
+        l = [(hom_ref, het, het, auto, 0, 2),
+         (hom_ref, hom_ref, het, auto, 0, 1),
+         (hom_ref, het, hom_ref, auto, 0, 1),
+         (het, het, het, auto, 1, 1),
+         (het, hom_ref, het, auto, 1, 0),
+         (het, het, hom_ref, auto, 1, 0),
+         (het, hom_var, het, auto, 0, 1),
+         (het, het, hom_var, auto, 0, 1),
+         (hom_var, het, het, auto, 2, 0),
+         (hom_var, het, hom_var, auto, 1, 0),
+         (hom_var, hom_var, het, auto, 1, 0),
+         (hom_ref, hom_ref, het, hemiX, 0, 1),
+         (hom_ref, hom_var, het, hemiX, 0, 1),
+         (hom_var, hom_ref, het, hemiX, 1, 0),
+         (hom_var, hom_var, het, hemiX, 1, 0)]
+
+        mapping = {Struct({'kid': v[0], 'dad': v[1], 'mom': v[2], 'ploidy': v[3]}): [v[4], v[5]] for v in l}
+
+        trio_matrix = trio_matrix.annotate_global('global.mapping', mapping, t)
+
+        trio_matrix = trio_matrix.annotate_variants_expr(
+            'va.category = if (v.isAutosomal() || v.inXPar() || v.inYPar()) 0 else if (v.inXNonPar()) 1 else -1')
+
+
+        s = '''
+            va.{name} = gs
+                .filter(g => va.category != 1 || !g.father.isHet())
+                .map(g =>
+                    let ploidy =
+                        if (va.category == 0) 2
+                        else if (va.category == -1) -1
+                        else if (sa.isFemale) 2
+                        else 1 in
+                        global.mapping.get(
+                            {{kid: g.proband.nNonRefAlleles(),
+                            dad: g.father.nNonRefAlleles(),
+                            mom: g.mother.nNonRefAlleles(),
+                            ploidy: ploidy}}
+                        )[{index}])
+                .sum()'''
+
+        per_variant_results = (
+            trio_matrix
+                .annotate_variants_expr([s.format(name='t', index=0), s.format(name='u', index=1)])
+                .variants_table()
+                .annotate('transmitted = va.t, untransmitted = va.u')
```

Comment:
are these supposed to correspond to 'nTransmitted' and 'nUntransmitted' in the docs?